### PR TITLE
♻️ Pipeline Builder code changes for nested pipeline config (distortion_correction) 

### PR DESCRIPTION
**Pull request by shnizzedy**
## Related to

Related to #1366 by @sgiavasis 

## Description

* ♻️ Updates calls to `c.`{configuration keys} in distortion_correction to nested versions in [default_pipeline.yml@`45796e`](https://github.com/FCP-INDI/C-PAC/blob/45796e02a091e98da9d65f882b02ea8aa8a10986/dev/docker_data/default_pipeline.yml)
* :rotating_light: Lints those same lines
* :goal_net: Update schema for the above changes

### Pipeline
* https://github.com/FCP-INDI/C-PAC/blob/6eadaed317a9b256a82f15bccee57a5e8de7514e/dev/docker_data/default_pipeline.yml#L756-L757 becomes https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L692-L702 (#1050) so I added https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/distortion_correction/distortion_correction.py#L409-L411 to the top of the function
* This block seems one level not-indented-far-enough: https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/distortion_correction/distortion_correction.py#L551-L563 If I'm reading the code correctly, `strat` is defined in this loop https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/distortion_correction/distortion_correction.py#L494 so the whole block in question here would only apply to the final iteration of that loop. Possibly related: #1072, #1131, #1242

### Schema
* Starting with (and including) schema changes from #1380
* There are a handful of mutually-exclusive keys that can be true here https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L727
   To enforce that mutual exclusivity: https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/pipeline/schema.py#L5-L22 https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/pipeline/schema.py#L162-L182  Effectively, the validator will exclusively accept any of the following for `c.functional_preproc['func_masking']['FSL-BET']`:
```python
[{'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': True,
  'reduce_bias': True,
  'robust': False,
  'padding': False,
  'remove_eyes': False,
  'surfaces': False},
 {'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': True,
  'robust': True,
  'reduce_bias': False,
  'padding': False,
  'remove_eyes': False,
  'surfaces': False},
 {'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': True,
  'padding': True,
  'reduce_bias': False,
  'robust': False,
  'remove_eyes': False,
  'surfaces': False},
 {'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': True,
  'remove_eyes': True,
  'reduce_bias': False,
  'robust': False,
  'padding': False,
  'surfaces': False},
 {'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': True,
  'surfaces': True,
  'reduce_bias': False,
  'robust': False,
  'padding': False,
  'remove_eyes': False},
 {'frac': float,
  'mesh_boolean': bool,
  'outline': bool,
  'radius': int,
  'skull': bool,
  'threshold': bool,
  'vertical_gradient': Range(min=-1, max=1, min_included=False, max_included=False, msg=None),
  'functional_mean_boolean': bool,
  'reduce_bias': False,
  'robust': False,
  'padding': False,
  'remove_eyes': False,
  'surfaces': False}]
```

I have 2 questions about this mutual exclusivity:
1. Am I reading correctly that `functional_mean_boolean` must be `True` for any of the mutually-exclusive options to be `True` https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L727-L729
2. What is `functional` referring to in the mutually-exclusive comment? These are the mutually-exclusive keys I see
   | old | new |
   | --- | --- |
   | https://github.com/FCP-INDI/C-PAC/blob/6eadaed317a9b256a82f15bccee57a5e8de7514e/dev/docker_data/default_pipeline.yml#L271-L272 https://github.com/FCP-INDI/C-PAC/blob/6eadaed317a9b256a82f15bccee57a5e8de7514e/dev/docker_data/default_pipeline.yml#L279-L288 https://github.com/FCP-INDI/C-PAC/blob/6eadaed317a9b256a82f15bccee57a5e8de7514e/dev/docker_data/default_pipeline.yml#L295-L296 | https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L740-L741 https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L746-L753 https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L758-L759 |

## Tests

```Python
from CPAC.pipeline.schema import schema
schema(config)
```

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `enh/nested_config` branch of the repository. 
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I updated the changelog.
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sgiavasis** [comment]
@HechengJin0 @SucyLi I think the `functional_mean_boolean` thing is new? What is the design behind that one? I see some comments in default_pipeline.yml on it but it doesn't fully make sense to me.

**Comment by HechengJin0** [comment]
Variations on default bet2 functionality (mutually exclusive options):
  (default)   just run bet2
  -R          robust brain centre estimation (iterates BET several times)
  -S          eye & optic nerve cleanup (can be useful in SIENA - disables -o option)
  -B          bias field & neck cleanup (can be useful in SIENA)
  -Z          improve BET if FOV is very small in Z (by temporarily padding end slices)
  -F          apply to 4D FMRI data (uses -f 0.3 and dilates brain mask slightly)
  -A          run bet2 and then betsurf to get additional skull and scalp surfaces (includes registrations)
  -A2 <T2>    as with -A, when also feeding in non-brain-extracted T2 (includes registrations)
https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/BET/UserGuide#Variations_on_default_bet2_functionality_.28mutually_exclusive_options.29:

need more clear description. 

1. for my understanding, if `functional_mean_boolean` is true, mutually-exclusive options (e.g. reduce_bias,robust,padding) should be false. 
2. `functional_mean_boolean` should be referring to `funcitonal` 
https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L729

**Comment by shnizzedy** [comment]
Thanks @HechengJin0!

If `functional_mean_boolean` is mutually exclusive with  `reduce_bias` , `robust`, `padding`, `remove_eyes`, and `surfaces`

What does the comment just above `functional_mean_boolean`
```Python
 # It must be 'on' if select 'bold_bet_reduce_bias','bold_bet_robust','bold_bet_padding','bold_bet_remove_eyes',or 'bold_bet_surfaces' on
functional_mean_boolean: Off
```
mean?

**Comment by sgiavasis** [comment]
I'm not entirely clear on what is being decided here:

https://github.com/FCP-INDI/C-PAC/blob/55ae58c004307294dae41ba4a60120ca444bf465/CPAC/func_preproc/func_preproc.py#L339-L353

It passes the `functional` flag to BET if `functional_mean_boolean` is false?

**Comment by shnizzedy** [comment]
Oh, I see! 

[`functional`](https://github.com/nipy/nipype/blob/b356ad16828afd4c63f862dfc72fa7e841863391/nipype/interfaces/fsl/preprocess.py#L114) is a boolean for `"apply to 4D fMRI data"`.

Here's the whole `nipype.interfaces.fsl.BET.help()`:

```Python
Wraps the executable command ``bet``.

FSL BET wrapper for skull stripping

For complete details, see the `BET Documentation.
<https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/BET/UserGuide>`_

Examples
--------
>>> from nipype.interfaces import fsl
>>> btr = fsl.BET()
>>> btr.inputs.in_file = 'structural.nii'
>>> btr.inputs.frac = 0.7
>>> btr.inputs.out_file = 'brain_anat.nii'
>>> btr.cmdline
'bet structural.nii brain_anat.nii -f 0.70'
>>> res = btr.run() # doctest: +SKIP

Inputs::

        [Mandatory]
        in_file: (a pathlike object or string representing an existing file)
                input file to skull strip
                argument: ``%s``, position: 0

        [Optional]
        out_file: (a pathlike object or string representing a file)
                name of output skull stripped image
                argument: ``%s``, position: 1
        outline: (a boolean)
                create surface outline image
                argument: ``-o``
        mask: (a boolean)
                create binary mask image
                argument: ``-m``
        skull: (a boolean)
                create skull image
                argument: ``-s``
        no_output: (a boolean)
                Don't generate segmented output
                argument: ``-n``
        frac: (a float)
                fractional intensity threshold
                argument: ``-f %.2f``
        vertical_gradient: (a float)
                vertical gradient in fractional intensity threshold (-1, 1)
                argument: ``-g %.2f``
        radius: (an integer)
                head radius
                argument: ``-r %d``
        center: (a list of at most 3 items which are an integer)
                center of gravity in voxels
                argument: ``-c %s``
        threshold: (a boolean)
                apply thresholding to segmented brain image and mask
                argument: ``-t``
        mesh: (a boolean)
                generate a vtk mesh brain surface
                argument: ``-e``
        robust: (a boolean)
                robust brain centre estimation (iterates BET several times)
                argument: ``-R``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        padding: (a boolean)
                improve BET if FOV is very small in Z (by temporarily padding end
                slices)
                argument: ``-Z``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        remove_eyes: (a boolean)
                eye & optic nerve cleanup (can be useful in SIENA)
                argument: ``-S``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        surfaces: (a boolean)
                run bet2 and then betsurf to get additional skull and scalp surfaces
                (includes registrations)
                argument: ``-A``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        t2_guided: (a pathlike object or string representing a file)
                as with creating surfaces, when also feeding in non-brain-extracted
                T2 (includes registrations)
                argument: ``-A2 %s``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        functional: (a boolean)
                apply to 4D fMRI data
                argument: ``-F``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        reduce_bias: (a boolean)
                bias field and neck cleanup
                argument: ``-B``
                mutually_exclusive: functional, reduce_bias, robust, padding,
                  remove_eyes, surfaces, t2_guided
        output_type: ('NIFTI' or 'NIFTI_PAIR' or 'NIFTI_GZ' or
                  'NIFTI_PAIR_GZ')
                FSL output type
        args: (a string)
                Additional parameters to the command
                argument: ``%s``
        environ: (a dictionary with keys which are a bytes or None or a value
                  of class 'str' and with values which are a bytes or None or a
                  value of class 'str', nipype default value: {})
                Environment variables

Outputs::

        out_file: (a pathlike object or string representing a file)
                path/name of skullstripped file (if generated)
        mask_file: (a pathlike object or string representing a file)
                path/name of binary brain mask (if generated)
        outline_file: (a pathlike object or string representing a file)
                path/name of outline file (if generated)
        meshfile: (a pathlike object or string representing a file)
                path/name of vtk mesh file (if generated)
        inskull_mask_file: (a pathlike object or string representing a file)
                path/name of inskull mask (if generated)
        inskull_mesh_file: (a pathlike object or string representing a file)
                path/name of inskull mesh outline (if generated)
        outskull_mask_file: (a pathlike object or string representing a file)
                path/name of outskull mask (if generated)
        outskull_mesh_file: (a pathlike object or string representing a file)
                path/name of outskull mesh outline (if generated)
        outskin_mask_file: (a pathlike object or string representing a file)
                path/name of outskin mask (if generated)
        outskin_mesh_file: (a pathlike object or string representing a file)
                path/name of outskin mesh outline (if generated)
        skull_mask_file: (a pathlike object or string representing a file)
                path/name of skull mask (if generated)
        skull_file: (a pathlike object or string representing a file)
                path/name of skull file (if generated)

References:
-----------
BibTeX('@article{JenkinsonBeckmannBehrensWoolrichSmith2012,author={M. Jenkinson, C.F. Beckmann, T.E. Behrens, M.W. Woolrich, and S.M. Smith},title={FSL},journal={NeuroImage},volume={62},pages={782-790},year={2012},}', key='JenkinsonBeckmannBehrensWoolrichSmith2012')
```

**Comment by shnizzedy** [comment]
So right now, we've got `functional_mean_boolean` deciding (`True`) to run `3dTstat` and then `fsl.BET` with any option except `functional=True` or (`False`) to run `fsl.BET` with `functional=True` and none of its mutually exclusive options set to `True` (`reduce_bias`, `robust`, `padding`, `remove_eyes`, `surfaces`)
functional_mean_boolean

**Comment by shnizzedy** [comment]
I think after merging this, we could clarify this step with a key `functional` making that key being `True` mutually exclusive with the other mutually-exclusive options like 
`functional` option

In that case, https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L727 would be pretty self-explanatory and https://github.com/FCP-INDI/C-PAC/blob/c3e44c0e31b02b85f92358065b7076554cfceb0b/dev/docker_data/default_pipeline.yml#L728 would be unnecessary because the reverse would be true.


Alternatively, we can just have a key like `mutually_exclusive_option` that takes a string or Nonetype to choose 0 or 1 of these options:

`mutually_exclusive_option` option

**Comment by shnizzedy** [comment]
I think the mutual exclusivity thing is refactored correctly to work with the nested-config as-is; if we want to do one of [the refactors suggested above](https://github.com/FCP-INDI/C-PAC/pull/1400#issuecomment-727067957) (or some other clarification of this step), I think that can be a separate PR.

File: CPAC/distortion_correction/distortion_correction.py [hunk 1]
```diff
@@ -498,19 +548,18 @@ def connect_distortion_correction(workflow, strat_list, c, diff, blip,
             workflow.connect(match_epi_fmaps_node, 'same_pe_epi',
                              blip_correct, 'inputspec.same_pe_epi')
 
-        if "None" in c.distortion_correction:
+        if False in c.functional_preproc['distortion_correction']['run']:
```
Comment (shnizzedy) [inline CPAC/distortion_correction/distortion_correction.py:551]: So the only [concern I have left](https://github.com/FCP-INDI/C-PAC/pull/1400#issue-519385702) before merging this PR into [`enh/nested_config`](https://github.com/FCP-INDI/C-PAC/tree/enh/nested_config) is 

> * This block seems one level not-indented-far-enough: https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/distortion_correction/distortion_correction.py#L551-L563 If I'm reading the code correctly, `strat` is defined in this loop https://github.com/FCP-INDI/C-PAC/blob/5e3581e04c1357af797f46482d75eaa368d3f84c/CPAC/distortion_correction/distortion_correction.py#L494 so the whole block in question here would only apply to the final iteration of that loop. Possibly related: #1072, #1131, #1242

**Comment by sgiavasis** [comment]
Either one of those refactorings would be more ideal. But I keep coming back to this and I think what is needed is to zoom out and ask why is this even an option in the first place? Looking at the entire pipeline.

Why would a user want to flag this?

@HechengJin0 , this was put into place 10 months ago (https://github.com/FCP-INDI/C-PAC/commit/7a515225c4759935b0ed1e52a00273eb092afafb )- do you remember what the impetus was?

Looks like the old default was to get the BOLD mask off the mean BOLD (3D): https://github.com/FCP-INDI/C-PAC/blob/fb5203edb8b909b557b16b348c4c1296973af330/CPAC/func_preproc/func_preproc.py#L70-L113

How does telling it to use the time series instead change things, why did we want that option available? I can't remember and I can't tell from looking at the code.

**Comment by HechengJin0** [comment]
I've checked v1.3.0, old default was to get the bold mask based on the`bet -F`functtional  https://github.com/FCP-INDI/C-PAC/blob/v1.3.0/CPAC/func_preproc/func_preproc.py#L50-L69. I think that the reason we set `functional_mean_boolean: off`. 
If I remember correctly, only rodent config set as `functional_mean_boolean: on`, https://github.com/FCP-INDI/C-PAC/issues/1107. Since rodent functional masking needs to set `bet -R` robust, which is conflict with -F functional.

**Review by sgiavasis — APPROVED** [review]
I'm finally folding this in, at least officially. I think that option needs to be changed, but let's deal with it in 1.8.1.

File: CPAC/distortion_correction/distortion_correction.py (hunk 1 above)
Comment (shnizzedy) [reply CPAC/distortion_correction/distortion_correction.py]: Resolved in #1423

**Pull request merged**